# Introduction à Python

## 📘 Syntaxe de base


`print()` est utilisé pour afficher du texte à la console.  
Les lignes débutant par `#` sont des commentaires et sont ignorées par Python.

In [ ]:
# TODO: Afficher la réponse à la grande question sur la vie, l'Univers et le reste
print(42)


42


### Les opérateurs

La liste des opérateurs mathématiques est identifiée dans ce tableau

Symbole|Description|Représentation mathématique|Exemple|Résultat
--|--|--|--|--
+|Addition|$x + y$|3 + 4|7
-|Soustraction|$x - y$|10 - 4|6
\*|Multiplication|$x \times y$|5 \* 3|15
/|Division|$x \div y$|15 / 2|7.5
**|Exposant|$x^y$|2 \*\* 3|8
//|Division entière|$\lfloor x \div y \rfloor$|15 // 2|7
%|Modulo|$x \ mod \ y$|15 % 2|1

In [ ]:
# TODO: Afficher les résultats avec print et les opérateurs
print(3 + 4)
print(10 - 4)

# Nb atomes dans l'univers
print(2**130)

7
6
1361129467683753853853498429727072845824


Les parenthèses peuvent être utilisées pour la priorité d'opération. Par exemple :
- $3 + 4 \times 8$
- $(3 + 4) \times 8$

In [ ]:
# TODO: Afficher les résultats sans et avec les parenthèses
print( 3 + 4 * 8 )
print( (3 + 4) * 8)
print( 7 / 3 )

35
56
2.3333333333333335


❗Attention à ce que le code soit aligné correctement sur la même colonne. Le code suivant
```python
print(3 + 4)
 print(4 + 5)
print(5 + 6)
```
donnera une erreur de syntaxe.

In [ ]:
# TODO: Exécuter le code pour pour obtenir une erreur de syntaxe.
#       Par la suite, corriger l'erreur.
print(3 + 4)
print(4 + 5)
print(5 + 6)

7
9
11


Les nombres décimaux doivent être écrits avec un `.` et leur résultat peut être approximatif.

In [ ]:
# TODO: Afficher une approximation de PI
print(3.141592654)

# TODO: Afficher l'addition de 0.1 et de 0.2.
print(0.1 + .2)

3.141592654
0.30000000000000004


### 📄 Exercices
Faire les exercices de la section **La syntaxe de base**

## 📘 Les variables

Les variables permettent de stocker une valeur afin de la réutiliser plus tard. Pour assigner la valeur `3` à la variable `x`, on utilise le symbole `=`.
```python
x = 3
```
On pourra par la suite afficher ou utiliser cette variable.
```python
print(x)
```

In [ ]:
# TODO: Assigner la valeur 3 à la variable x et l'afficher.
x = 5
print(x)

5


### Nommer une variable

Le nom des variables peuvent être des lettres ou des mots. Ceux-ci doivent être significatifs pour aider à la compréhension du programme. Elles peuvent:
- être de n'importe quelle longueur;
- contenir des lettres et des chiffres;
- contenir des lettres minuscules et majuscules.

Cependant, elles doivent:
- débuter par une lettre;
- ne pas contenir d'espaces;
- ne pas contenir de caractères spéciaux autre que le souligné `_`.

Par exemple, l'exercice 5 passe de:
```python
print(-4.9 * 4**2 + 10 * 4 + 50)
```
à:
```python
t = 4
print(-4.9 * t**2 + 10 * t + 50)
```
ou même à:
```python
temps = 4
print(-4.9 * temps**2 + 10 * temps + 50)
```

Essayez avec l'exercice 6 en remplaçant la valeur de $x$ par une variable dans le calcul
$$f(x) = \frac{x^2 -2}{3(3x + 2)²}$$
lorsque $x = -2$

In [ ]:
# TODO: Utiliser des variables dans le calcul.
x = 5
print( (x**2 - 2) / (3 * (3 * x + 2) ** 2) )

0.0265282583621684


#### Constantes
Certaines variables ne devraient pas changer au cours d'un programme. On les appelle **constantes**. Une convention est de nommer la variable en n'utilisant que des majuscules. Par exemple, on pourrait définir le taux de la TPS et de la TVQ comme suit:
```python
TAUX_TPS = 0.05
TAUX_TVQ = 0.09975
```

### Les types de variables

Plusieurs types de variables sont disponibles en Python.

Nom (terme en programmation)|Description|Exemple
--|--|--
Entier (*integer*, *int*)|Nombre entier, positif ou négatif|`42`
Nombre décimal (*float*)|Nombre à virgule|`12.34`
Booléen (*boolean*, *bool*)|Vrai ou Faux|`True`, `False`
Texte (*string*, *str*)|Chaîne de caractères, texte|`"Bonjour le monde!"`<br>`'Bonjour le monde!'`
Liste ou tableau (*list*, *array*)|Plusieurs valeurs dans une variable|`[4, 8, 72, 45]`
❗Un texte (*string*) peut être entouré de guillements `"` ou d'apostrophes `'`.

In [ ]:
print(42)                   # Nombre entier
print(12.34)                # Nombre à virgule
print(True)                 # Booléen
print("Bonjour le monde!")  # Texte
print('Bonjour Lévis!')     # Autre texte
print([4, 8, 72, 45])       # Liste

42
12.34
True
Bonjour le monde!
Bonjour Lévis!
[4, 8, 72, 45]


Grâce à l'opérateur `+`, il sera même possible de **concaténer** du texte ensemble:
```python
ville = "Lévis"
print("Nous sommes à " + ville)
```

In [ ]:
ville = "Gatineau"
print("Nous sommes à " + ville)

Nous sommes à Gatineau


### Conversion de type

Cependant, il n'est pas possible d'additionner du texte et des nombres sans convertir les valeurs. Le code suivant donnera une erreur:
```python
age = 2024 - 1972
print("Mon age est " + age)
```
car `age` est un nombre qu'on veut concaténer (`+`) avec du texte.

In [ ]:
age = 2024 - 1972
print("Mon age est " + age)

TypeError: can only concatenate str (not "int") to str

Pour convertir une valeur vers un autre type, plusieurs fonctions sont à notre disposition.

Fonction|Description|Exemple --> Résultat
--|--|--
`int()`|Converti en entier|`int("36")` --> `36`
`float()`|Converti en décimal|`float(36)` --> `36.0`
`str()`|Converti en texte|`str(36)` --> `"36"`
`bool()`|Converti en booléen|`bool(36)` --> `True`

On pourra alors modifier le code précédent pour convertir la variable `age` en texte avant de l'afficher:
```python
age = 2024 - 1972
print("Mon age est " + int(age))
```

In [ ]:
# TODO: Corriger le code suivant en convertissant la variable "age".
age = 2024 - 1972
print("Mon age est " + str(age))

Mon age est 52


### Un premier programme interractif avec fonction `input()`

Jusqu'à présent, pour modifier le résultat du programme, on doit modifier le code. Avec la fonction `input()`, il est possible de demander de l'information à l'utilisateur. Le résultat entré par l'utilisateur pourra alors être assigné à une variable.

❗Le type de variable retourné par `input()` est du texte, un *string*.

```python
nom = input("Quel est votre nom? ")
print("Bonjour " + nom)
```

In [ ]:
# TODO: Écrire un programme qui demande votre nom et votre ville.
#       Afficher par la suite une formule de signature.
#       Ex. Signé par LE_NOM à LA_VILLE
nom = input("Quel est votre nom? ")
ville = input("Quelle est la ville? ")
print("Signé par " + nom + " à " + ville)

Quel est votre nom? Sébastien
Quelle est la ville? Lévis
Signé par Sébastien à Lévis


Pourrait-on écrire un programme mathématique plus intéressant? Essayons d'écrire un programme qui multiplie par `5` un nombre demandé par `input()`. Tester le programme avec $8 \times 5$.

In [ ]:
# TODO: 1. Demander le nombre à multiplier par 5.
nombre = float(input("Quel est le nombre? "))

# TODO: 2. Afficher le résultat de la multiplication.
print(5 * nombre)

Quel est le nombre? 8
40.0


Selon comment le programme a été écrit, deux résultats pourraient être obtenus:
1. 40
2. 88888

Si on se souvient du rappel [plus haut](#scrollTo=Y04Xv1rPpEO5&line=1&uniqifier=1), le type de variable retourné par `input()` est un *string*. Il est possible que, sans le vouloir, le calcul effectué ait été:
```python
print("8" * 5)
```
ou en d'autres mots, multiplier une chaîne de caractère 5 fois, d'où le résultat 88888.

### Quelques astuces avec `print()`

La fonction `print()` est très utile pour afficher du texte. Cependant, il peut être difficile de voir le résultat final lorsque le texte est très long. Par exemple:
```python
cote_a = 3
cote_b = 4
cote_c = (cote_a**2 + cote_b**2) ** (1/2)
print("L'hypoténuse du triangle " + str(cote_a) + ", " + str(cote_b) + " est de " + str(cote_c) + ".")
```

Bien que fonctionnel, ce n'est pas très élégant. On peut utiliser ce qu'on appelle des *f-string* pour rendre le tout plus facile à lire. Pour utiliser cette fonctionnalité, on ajoute le caractère `f` en avant de la chaîne de caractère et on entoure les variables par des acollades `{` `}`.
```python
cote_a = 3
cote_b = 4
cote_c = (cote_a**2 + cote_b**2) ** (1/2)
print(f"L'hypoténuse du triangle {cote_a}, {cote_b} est de {cote_c}.")
```

In [ ]:
# TODO: Utiliser des f-string au lieu de la concaténation.
prenom = input("Quel est votre prénom? ")
nom = input("Bonjour " + prenom + ", quel est votre nom? ")
print("Bonjour " + prenom + " " + nom + ", heureux de faire votre connaissance.")

Quel est votre prénom? Sebastien
Bonjour Sebastien, quel est votre nom? Huneault
Bonjour Sebastien Huneault, heureux de faire votre connaissance.


❗Avec des *f-string*, il est également possible d'afficher facilement les nombres en notation scientifique ou avec une précision décimale.
```python
x = 4.176**10
print(f"{x}")       # 1612890.5595595774
print(f"{x:e}")     # 1.612891e+06
print(f"{x:.2e}")   # 1.61e+06
print(f"{x:.2f}")   # 1612890.56
```


Une autre astuce permet d'avoir plusieurs instructions `print()` qui s'affichent sur la même ligne. Normalement, le code
```python
print("Sébastien")
print("Olivier")
print("Charles")
```
donnerait
```
Sébastien
Olivier
Charles
```
Mais en ajoutant un paramètre `end=""`
```python
print("Sébastien", end="")
print("Olivier", end="")
print("Charles", end="")
```
le résultat change pour
```
SébastienOlivierCharles
```

In [ ]:
x = input("Quel est le nombre? ")
x = int(x)
x = x + 5
print(x)

Quel est le nombre? 2
7


### 📄 Exercices
Faire les exercices de la section **Les variables, `print()` et `input()`**

## 📘 Les fonctions

### Les fonctions de base

En plus des fonctions `print()` et `input()`, Python quelques fonctions pour effectuer des calculs de base.

Fonction|Description|Exemple
--|--|--
`abs(nombre)`|Retourne la valeur absolue d'un nombre|`abs(3)` retourne `3`<br>`abs(-7)` retourne `7`
`min(nombre1, nombre2, ...)`<br>`max(nombre1, nombre2, ...)`|Retourne la valeur minimale (ou maximale) de plusieurs nombres|`min(-3, 5, 7)` retourne `-3`<br>`max(-1, 3, 9, 4)` retourne `9`
`sum(nombre1, nombre2, ...)`|Retourne la somme des nombres|`sum(2, -3, 4)` donne `3`
`round(nombre, [précision])`|Retourne le nombre arrondi à l'entier le plus près. Si le paramètre `précision` (optionnel) est spécifié,<br>retourne le nombre de chiffres après la virgule souhaité.|`round(3.14159)` retourne `3`<br>`round(3.14159, 4)` retourne `3.1416`<br>`round(3.5)` retourne `4`

Ces fonctions sont déjà intégrées dans Python et peuvent être utilisées immédiatement.
```python
print(abs(-7))            # Affiche 7
print(sum(1, 2, 3, 4, 5)) # Affiche 15
```

### Les fonctions mathématiques

En plus des fonctions intégrées, Python possède une panoplie de modules pour étendre ses capacités. Avant de pouvoir utiliser des modules et les fonctions qu'ils contiennent, ils doivent être **importées**. Le module `math` possède plusieurs fonctions trigonométriques en plus de constantes très utiles. Quelques fonctions utiles:
- `math.sin()`, `math.asin()`
- `math.cos()`, `math.acos()`
- `math.tan()`, `math.atan()`
- `math.pi`, `math.e`, `math.tau`

Pour importer le module `math` et pouvoir l'utiliser, on écrit la ligne suivante:
```python
import math
```
Seulement après on pourra utiliser ses fonctions:
```python
import math
print(math.sin(math.pi / 2))
```

In [ ]:
# TODO: Utiliser le module math pour la valeur de pi.
import math
PI = math.pi

rayon = float(input("Quel est le rayon? "))
aire = PI * rayon**2
print(f"L'aire du cercle est de {aire:.2f}")

Quel est le rayon? 1
L'aire du cercle est de 3.14


### Les fonctions personnalisées

#### Définir et utiliser une fonction

Les fonctions en Python peuvent ressembler à des petits programmes réutilisables. Une fois définie, elles pourront être appelées pour effectuer les opérations et possiblement retourner un résultat. Par exemple, on pourrait créer une fonction qui calcule l'hypoténuse d'un triangle rectangle comme ceci:
```python
def calculer_hypotenuse(cote_a, cote_b):
  hypotenuse = (cote_a**2 + cote_b**2) ** 0.5
  return hypotenuse
```
Par la suite, on pourrait utiliser la fonction pour obtenir l'hypoténuse en **appelant** la fonction `calculer_hypotenuse()` avec des valeurs:
```python
cote_c = calculer_hypotenuse(3, 4)
print(f"Le triangle 3, 4 a une hypoténuse de {cote_c}")
```

❗Les fonctions doivent être définies avant d'être utilisées. Ainsi, le code suivant donnerait une erreur:
```python
cote_c = calculer_hypotenuse(3, 4)
print(f"Le triangle 3, 4 a une hypoténuse de {cote_c}")

def calculer_hypotenuse(cote_a, cote_b):
  hypotenuse = (cote_a**2 + cote_b**2) ** 0.5
  return hypotenuse
```
De plus, l'indentation du code (les espaces en début de lignes) sont **très** importantes. Ces espaces indiquent au langage Python le contenu de la fonction. La touche `tabulation` peut être utilisée pour obtenir une indentation uniforme.

In [ ]:
def calculer_hypotenuse(cote_a, cote_b):
  hypotenuse = (cote_a**2 + cote_b**2) ** 0.5
  return hypotenuse
cote_c = calculer_hypotenuse(3, 4)
print(f"Le triangle 3, 4 a une hypoténuse de {cote_c}")

Le triangle 3, 4 a une hypoténuse de 5.0


In [ ]:
# TODO: Terminer le programme en remplaçant le calcul par une fonction.
# TODO: Demander la température à l'utilisateur.
# DÉFI: Demander la température à l'utilisateur et accepter les nombres décimaux.
def celsius_en_fahrenheit(temp):
  return temp * 9 / 5 + 32

def fahrenheit_en_celsius(temp):
  return (temp - 32) / 9 * 5

temperature = float(input("Quelle est la temperature? "))
fahrenheit = celsius_en_fahrenheit(temperature)
celsius = fahrenheit_en_celsius(temperature)

print(f"{temperature} celsius = {fahrenheit:.2f} fahrenheit")
print(f"{temperature} fahrenheit = {celsius:.2f} celsius")

Quelle est la temperature? 33.7
33.7 celsius = 92.66 fahrenheit
33.7 fahrenheit = 0.94 celsius


#### Valeurs de retour multiples

Une fonction peut retourner plus d'une valeur à la fois. Cette fonctionnalité peut s'avérer très pratique pour la fonction quadratique utilisée précédemment. La fonction n'a qu'à retourner plusieurs valeurs séparées par une virgule:
```python
def quadratique(a, b, c):
  x0 = ((-b) + (b**2 - 4*a*c)**.5) / (2*a)
  x1 = ((-b) - (b**2 - 4*a*c)**.5) / (2*a)
  return x0, x1
```
Par la suite, la fonction pourrait être utilisée comme suit:
```python
x0, x1 = quadratique(-2, 3, 10)
print(f"Les valeurs sont {x0} et {x1}")
```

❗Les fonctions ne sont pas obligées d'avoir de paramètres ni de retourner de valeur, mais elles deviennent moins pratiques:
```python
def salutations():
  nom = input("Quel est votre nom? ")
  print(f"Bonjour {nom}!")

salutations()
```

In [ ]:
def salutations():
  nom = input("Quel est votre nom? ")
  print(f"Bonjour {nom}!")

salutations()

Quel est votre nom? Sebastien
Bonjour Sebastien!


### 📄 Exercices
Faire les exercices de la section **Les fonctions**

## 📘 Le traitement conditionnel

Le traitement conditionnel permet d'exécuter certaines instructions selon une condition donnée. On utilise le mot-clé `if` suivi d'une condition à évaluer. Selon si le résultat est vrai (`True`) ou faux (`False`), certaines instructions pourraient s'exécuter ou non. Un exemple simple est de vérifier si un nombre est positif ou négatif:
```python
nombre = float(input("Entrer un nombre: "))
if nombre > 0:
  print(f"Le nombre {nombre} est positif")
```
Si le nombre `5` est entré, la phrase "Le nombre 5 est positif" s'affichera. Cependant, si le nombre `-3` est entré, rien ne sera affiché.

In [ ]:
nombre = float(input("Entrer un nombre: "))
if nombre > 0:
  print(f"Le nombre {nombre} est positif")

Entrer un nombre: -5


### Les opérateurs de comparaison

Une condition suit toujours le même format: `<valeur1> <operateur> <valeur2>`. Dans l'exemple précédent, `valeur1` était le nombre entré, `valeur2` était le nombre zéro et l'opérateur, le symbole `>`, soit *plus grand que*. Ce trio constitue une **expression** qui sera évaluée au final en un booléen vrai (`True`) ou faux (`False`). Si l'expression est vraie, le code à l'intérieur de la condition (visible par l'indentation) sera exécutée. Si l'expression est fausse, le code sous l'instruction `else` (si applicable) sera exécuté.

Les opérateurs suivants sont disponibles pour comparer des valeurs:

Symbole|Description|Exemple `True`|Exemple `False`
--|--|--|--
`==`|Équivalent ou égal à|`8 == 8`|`8 == 2`
`!=`|Différent de|`5 != 2`|`5 != 5`
`>`|Plus grand que|`8 > 5`|`8 > 12`<br>`12 > 12`
`>=`|Plus grand ou égal à|`8 >= 5`<br>`12 >= 12`|`8 >= 15`
`<`|Plus petit que|`5 < 8`|`12 < 8`<br>`12 < 12`
`<=`|Plus petit ou égal à|`5 <= 8`<br>`12 <= 12`|`15 <= 8`


❗La section sur les [opérateurs](#scrollTo=LbW-Pn7AVMML&line=1&uniqifier=1) pourrait être utile pour certaines expressions.

In [ ]:
# TODO: Remplacer les expressions dans les conditions pour rendre le programme fonctionnel
nombre = float(input("Entrer un nombre: "))

if nombre > 0:
  print(f"Le nombre {nombre} est positif")

if nombre < 0:
  print(f"Le nombre {nombre} est négatif")

if nombre >= 0:
  print(f"Le nombre {nombre} est positif ou zéro")

if nombre % 2 == 0:
  print(f"Le nombre {nombre} est pair")

if nombre % 2 != 0:
  print(f"Le nombre {nombre} est impair")

Entrer un nombre: 6
Le nombre 6.0 est positif
Le nombre 6.0 est positif ou zéro
Le nombre 6.0 est pair


### Algèbre de Boole

#### Instruction `and`

Il pourrait être utile de vérifier plusieurs conditions avant d'exécuter des instructions. Par exemple, si on voulait vérifier si aujourd'hui est notre anniversaire, il faudrait effectuer 2 vérifications:
1. Est-ce le bon mois ?
1. Est-ce le bon jour ?

On peut utiliser l'opérateur **and** pour vérifier deux expressions en une seule commande:
```python
fete_mois = 1
fete_jour = 15
auj_mois = 1
auj_jour = 11

if fete_mois == auj_mois and fete_jour == auj_jour:
  print("C'est ma fête!")
```
Pour déterminer si l'expression complète sera vraie ou fausse, on doit évaluer chacune des expressions indépendemment:
- `fete_mois == auj_mois` => `True`
- `auj_mois == auj_jour`  => `False`

L'expression simplifiée alors `if True and False` et sera à nouveau évaluée pour obtenir la réponse finale.

Pour déterminer le résultat booléen avec l'instruction `and`, on peut consulter la table de vérité suivante:

Valeur1|Valeur2|Résultat
--|--|--
`False`|`False`|`False`
`False`|`True`|`False`
`True`|`False`|`False`
**`True`**|**`True`**|**`True`**

Pour l'opérateur `and`, le code sous-jacent sera exécuté seulement lorsque les deux valeurs sont à `True`.

In [ ]:
# TODO: Transformer les conditions imbriquées dans une seule à l'aide de l'instruction and
couleur = "jaune"
fruit = "banane"

if couleur == "jaune" and fruit == "banane":
  print("Ce fruit est prêt à être mangé!")

if couleur == "jaune":
  if fruit == "banane":
    print("Ce fruit est prêt à être mangé!")


annee = 2020
prix = 15000

if prix <= 18000 and annee >= 2019:
  print("Cette voiture m'intéresse.")

if prix <= 18000:
  if annee >= 2019:
    print("Cette voiture m'intéresse.")



Ce fruit est prêt à être mangé!
Cette voiture m'intéresse.


#### Instruction `or`

L'instruction `or`, quant à lui, permet d'exécuter du code lorsque seulement l'une des deux expressions est vraie. Cela permet, par exemple, de déterminer si nous sommes la fin de semaine:

```python
jour = "Samedi"

if jour == "Samedi" or jour == "Dimanche":
  print("C'est la fin de semaine!")
```

Similaire au `and`, on utilisera le même processus pour déterminer si l'expression finale est `True` ou `False`:
- `jour == "Samedi"` => `True`
- `jour == "Dimanche"` => `False`

L'expression finale sera donc `if True or False` et pourra à nouveau être réduite grace à la table de vérité suivante:

Valeur1|Valeur2|Résultat
--|--|--
`False`|`False`|`False`
`False`|**`True`**|**`True`**
**`True`**|`False`|**`True`**
**`True`**|**`True`**|**`True`**

Pour l'opérateur `or`, le code sous-jacent sera exécuté lorsqu'une seule des valeurs sont à `True`.

#### Combinaison de `and` et de `or`

On peut ajouter autant de `and` et de `or` que souhaité. Cependant, il faut faire attention à la priorité des opérations. Regardons le code suivant:
```python
age = 15
forfait = False
if age < 18 or age > 25 and forfait == True:
  print("Vous avez un rabais")
```
Lorsque le code sera exécuté, est-ce que la personne aura droit à un rabais?

In [ ]:
# TODO: Exécuter le code pour confirmer sa réponse
age = 15
forfait = True
if age < 18 or age > 25 and forfait == True:
  print("Vous avez un rabais")

Vous avez un rabais


En ce qui a trait à la priorité des opérations, on peut comparer l'instruction `and` à la multiplication et l'instruction `or`, à l'addition. Au final, l'expression pourrait se lire ainsi:
```python
if (age < 18) or (age > 25 and forfait == True)
       \/            \/             \/
if (  True  ) or ( False   and     True       )
       \/                   \/
if (  True  ) or (         False              )
              \/
if           True
```
Toute personne ayant moins de 18 ans aura droit à un rabais, ainsi qu'un personne agée de plus de 25 ans avec un forfait.

In [ ]:
# TODO: Modifier pour que le rabais s'applique aux personnes ayant un forfait, peu
#       importe l'age, ainsi qu'aux gens de moins de 18 ans ou de plus de 25 ans.
age = 15
forfait = False
if (age < 18 or age > 25) or (forfait == True):
  print("Vous avez un rabais")

Vous avez un rabais


❗Astuce: Lorsqu'un expression est comparée à `True`, on peut raccourcir l'expression pour n'utiliser que la valeur. Ainsi l'expression `if forfait == True:` peut être remplacée par `if forfait:`. Également, si l'expression est comparée à `False`, elle peut être remplacée par `if not forfait:`

#### Instruction `not`

Finalement, l'instruction `not` peut être utilisée devant une expression pour inverser sa valeur. Une expression résultant à `True` deviendra `False` et vice-versa. Ainsi, une autre façon de savoir si une personne a moins de 18 ans est de vérifier si elle **n'a pas** 18 ans ou plus:
```python
age = 15
if not age >= 18:
  print("J'ai moins de 18 ans")
```

In [ ]:
# TODO: Ajouter le mot-clé not pour obtenir le même résultat
age = 65
if not age < 18:
  print("J'ai le droit de vote")

est_sportif = True
if not est_sportif != True:
  print("Je suis sportif")

if age >= 65 and est_sportif == True:
  print("Je suis un sportif à la retraite")

if age < 18 and est_sportif == False:
  print("Je suis un jeune pas très sportif")

J'ai le droit de vote
Je suis sportif
Je suis un sportir à la retraite


### Les mots-clés `else` et `elif`

Le code affiché plus haut permettait d'indiquer si un nombre était positif:
```python
nombre = float(input("Entrer un nombre: "))
if nombre > 0:
  print(f"Le nombre {nombre} est positif")
```
Cependant, il ne permettait pas d'indiquer une réponse différente en cas de négation. Le mot-clé `else` vient résoudre ce problème:
```python
nombre = float(input("Entrer un nombre: "))
if nombre > 0:
  print(f"Le nombre {nombre} est positif")
else:
  print(f"Le nombre {nombre} est négatif")
```
Si le nombre est **plus grand que** que zéro, un message positif sera affiché. Dans **tous les autres** cas, un message négatif sera affiché. Ce sera soit l'un, soit l'autre.

In [ ]:
# TODO: Le message négatif est érroné et doit être corrigé.
nombre = float(input("Entrer un nombre: "))
if nombre >= 0:
  print(f"Le nombre {nombre} est positif")
else:
  print(f"Le nombre {nombre} est négatif")

Entrer un nombre: 0
Le nombre 0.0 est positif
Le nombre 0.0 est négatif


À l'aide des instructions `if` et `else`, il est possible de créer un embranchement de code à exécuter. Soit un bloc de code sera exécuté, soit un second. Deux choix. Le mot-clé `elif` permet de créer autant embranchements supplémentaires souhaité.

L'exemple précédent pourra alors être encore une fois bonifiée pour inclure le zéro dans les réponses:
```python
nombre = float(input("Entrer un nombre: "))
if nombre > 0:
  print(f"Le nombre {nombre} est positif")
elif nombre == 0:
  print(f"Le nombre {nombre} est zéro")
else:
  print(f"Le nombre {nombre} est négatif")
```

❗Attention! Seulement un seul bloc de code sera exécuté, même si plusieurs `elif` pourrait être vrai. Le premier sera alors choisi. Ainsi, le code suivant ne fonctionnerait pas comme prévu pour les nombres plus grand que 20:
```python
nombre = 25
if nombre > 0:
  print(f"Le nombre {nombre} est positif et moins grand que 20")
elif nombre > 20:
  print(f"Le nombre {nombre} est plus grand que 20")
else:
  print(f"Le nombre {nombre} est négatif")
```

In [ ]:
# TODO: Corriger le code pour obtenir le résultat attendu
nombre = 25
if nombre > 0 and nombre < 20:
  print(f"Le nombre {nombre} est positif et moins grand que 20")
elif nombre > 20:
  print(f"Le nombre {nombre} est plus grand que 20")
else:
  print(f"Le nombre {nombre} est négatif")

Le nombre 25 est plus grand que 20


### 📄 Exercices
Faire les exercices de la section **Le traitement conditionnel**

## 📘 La boucle `for`

Les boucles permettent d'exécuter une ou plusieurs instructions plusieurs fois. Elles peuvent être pratique pour compter, chercher, calculer des sommes, générer des listes de nombres, etc. Il existe deux types de boucles:
- la boucle `for`, utile lorsque le nombre de répétition est connu, et
- la boucle `while`, utile lorsque le nombre de répétition est inconnu.

La boucle `for` exécute une ou plusieurs instructions un nombre de fois bien précis.
```python
for i in range(10):
  print(f"Le nombre est {i}")
```
Plusieurs éléments se trouvent dans ce bout de code:
- la variable `i` est une variable qui changera de valeur à chaque itération;
- l'instruction `range(10)` permet de générer une liste de nombre entre 0 et 9.

Lorsqu'exécuté, l'instruction `print()` sera donc exécutée 10 fois, tout en changeant la valeur de `i` à chaque fois.

In [ ]:
# TODO: Répéter 20 fois au lieu de 10
print("** DÉBUT DE LA BOUCLE")
for i in range(20):
  print(f"Le nombre est {i}")

print("** FIN DE LA BOUCLE")

** DÉBUT DE LA BOUCLE
Le nombre est 0
Le nombre est 1
Le nombre est 2
Le nombre est 3
Le nombre est 4
Le nombre est 5
Le nombre est 6
Le nombre est 7
Le nombre est 8
Le nombre est 9
Le nombre est 10
Le nombre est 11
Le nombre est 12
Le nombre est 13
Le nombre est 14
Le nombre est 15
Le nombre est 16
Le nombre est 17
Le nombre est 18
Le nombre est 19
** FIN DE LA BOUCLE


L'instruction `range()` peut être modifiée pour
- débuter à un nombre différent de zéro: `range(4, 20)`
- changer le saut à chaque itération: `range(0, 20, 2)`

In [ ]:
# TODO: Afficher seulement les nombres entre 10 et 30 inclusivement
for i in range(10, 31):
  print(f"Le nombre est {i}")

In [ ]:
# TODO: Afficher seulement les nombres pairs entre 10 et 20 inclusivement
for i in range(10, 21, 2):
  print(f"Le nombre est {i}")

Le nombre est 10
Le nombre est 12
Le nombre est 14
Le nombre est 16
Le nombre est 18
Le nombre est 20


In [ ]:
# TODO: Afficher seulement les nombres impairs entre 7 et 19 inclusivement
for i in range(7, 20, 2):
  print(f"Le nombre est {i}")

Le nombre est 7
Le nombre est 9
Le nombre est 11
Le nombre est 13
Le nombre est 15
Le nombre est 17
Le nombre est 19


À l'intérieur d'une boucle, toutes les instructions apprises précédemment sont disponibles, incluant les conditions. Cela permet, par exemple, de trouver tous les nombres divisibles par $n$ entre une plage de nombres:
```python
n = 7
for i in range(50):
  if i % n == 0:
    print(f"Le nombre {i} est divisible par {n}")
```

In [ ]:
# TODO: Modifier le code pour:
#       - Exclure le zéro
#       - Afficher les nombres divisibles par 5 ou 7
for i in range(1, 50):
  if i % 7 == 0:
    print(f"Le nombre {i} est divisible par 7")
  if i % 5 == 0:
    print(f"Le nombre {i} est divisible par 5")

Une fonction très populaire à écrire serait de déterminer si un nombre est un nombre premier ou non. Dès qu'un diviseur est trouvé, le nombre n'est pas premier (`return False`). Autrement, quand tous les nombres ont été essayés sans trouver de diviseur, le nombre est premier (`return True`).
```python
def est_premier(nombre):
  for i in range(2, nombre):
    if nombre % i == 0:
      return False
  return True
```
On pourra alors trouver tous les nombres premiers entre 2 et 30 à l'aide d'une autre boucle:
```python
for i in range(2, 30):
  if est_premier(i):
    print(f"{i} est premier")
  else:
    print(f"{i}")
```
❗ Rappel: L'expression `if est_premier(i):` est l'équivalent de `if est_premier(i) == True:`

In [ ]:
# TODO: Essayer le code.
# TODO: (optionnel) Rendre la fonction est_premier() plus efficace.
def est_premier(nombre):
  for i in range(2, nombre):
    if nombre % i == 0:
      return False
  return True

for i in range(2, 5000):
  if est_premier(i) == True:
    print(f"{i} est premier")
  # else:
  #   print(f"{i}")

### Calculer une somme

Calculer une somme de nombres est maintenant chose facile grâce à la boucle. En réutilisant une variable, on pourrait manuellement faire la somme des 5 premiers nombres:
```python
total = 0
total = total + 1
total = total + 2
total = total + 3
total = total + 4
total = total + 5
print(total)    # Affiche 15
```
En ajoutant une boucle, il sera désormais possible de faire une somme bien plus complexe et de bien plus grands nombres:
```python
total = 0
for i in range(50):
  total = total + i
print(total)    # Affiche 1225
```

Finalement, en incluant le traitement conditionnel, il est possible de faire des calculs difficiles à faire à la main.

In [ ]:
# TODO: Calculer la somme des nombres premiers plus petits que 1000.
# TODO: (optionnel) Soustraire du total la somme des nombres divisibles par 7 plus petit que 572.
def est_premier(nombre):
  for i in range(2, nombre):
    if nombre % i == 0:
      return False
  return True

total = 0
total_div_7 = 0
for i in range(1000):
  if est_premier(i):
    total = total + i

  # if i % 7 == 0 and i < 572:
  #   total_div_7 = total_div_7 + i

for i in range(572):
  if i % 7 == 0:
    total_div_7 = total_div_7 + i

print(f"Total: {total}")
print(f"Total: {total - total_div_7}")

Total: 76128
Total: 52881


### Les instructions `break` et `continue`

L'instruction `break` permet de sortir complètement d'une boucle comme si elle était terminée. Le code
```python
for i in range(10):
  if i == 5:
    break
  print(f"Nombre {i}")
```
affichera les nombres entre 0 et 5. Rendu au nombre 6, l'instruction `break` sera exécutée et le code sera terminé.

In [ ]:
# TODO: Essayer le code
print("** DÉBUT DE LA BOUCLE")
for i in range(10):
  if i == 5:
    break
  print(f"Nombre {i}")
print("** FIN DE LA BOUCLE")

** DÉBUT DE LA BOUCLE
Nombre 0
Nombre 1
Nombre 2
Nombre 3
Nombre 4
** FIN DE LA BOUCLE


L'instruction `continue` permet de continuer immédiatement la boucle en ignorant les instructions sous elle. Ainsi, le code
```python
for i in range(10):
  if i == 5:
    continue
  print(f"Nombre {i}")
```
affichera les nombres entre 0 et 9, mais passera par dessus le nombre 5.

In [ ]:
# TODO: Essayer le code
print("** DÉBUT DE LA BOUCLE")
for i in range(10):
  if i == 5:
    continue
  print(f"Nombre {i}")
print("** FIN DE LA BOUCLE")

** DÉBUT DE LA BOUCLE
Nombre 0
Nombre 1
Nombre 2
Nombre 3
Nombre 4
Nombre 6
Nombre 7
Nombre 8
Nombre 9
** FIN DE LA BOUCLE


### Les boucles imbriquées


Plusieurs boucles peuvent être imbriquées l'une dans l'autre. Ceci permet d'effectuer une boucle $m$ fois. On pourrait s'en servir, par exemple, pour afficher une table de multiplication.

In [ ]:
# TODO: Changer le modificateur ^4 (ligne 7) pour un autre nombre et observer le résultat
# TODO: Changer les valeurs de range() de i et observer le résultat
# TODO: Changer les valeurs de range() de j et observer le résultat
# TODO: (optionnel) Ne pas afficher les valeurs carrés
for i in range(1, 11):
  for j in range(1, 11):
    total = i * j
    print(f"{total:>4}", end="")
  print()

   1   2   3   4   5   6   7   8   9  10
   2   4   6   8  10  12  14  16  18  20
   3   6   9  12  15  18  21  24  27  30
   4   8  12  16  20  24  28  32  36  40
   5  10  15  20  25  30  35  40  45  50
   6  12  18  24  30  36  42  48  54  60
   7  14  21  28  35  42  49  56  63  70
   8  16  24  32  40  48  56  64  72  80
   9  18  27  36  45  54  63  72  81  90
  10  20  30  40  50  60  70  80  90 100


❗Astuce! Le modificateur `{total:^4}` permet de centrer le nombre dans un espace de 4.

### 📄 Exercices
Faire les exercices de la section **La boucle `for`**

## 📘 La boucle `while`

Comme la boucle `for`, la boucle `while` permet d'effectuer une ou plusieurs instructions. La différence principale est qu'elle est utile lorsque la quantité d'itération est inconnue.

La syntaxe pour la boucle `while` est un mixte entre la boucle `for` et un conditionnel `if`.

```python
nombre = int(input("Entrer un nombre: "))
while nombre > 0:
  print(f"Décompte: {nombre}")
  nombre = nombre - 1
```

La boucle `while` n'a aucune idée du nombre d'itération. Cependant, elle continuera tant que le `nombre` est plus grand que zéro.

❗Attention! Étant donné que la boucle n'a pas de quantité d'itération définie d'avance, il faut faire attention aux boucles infinies!

In [ ]:
# TODO: Corriger le code pour enlever la boucle infinie. Appuyer sur STOP pour arrêter.
nombre = int(input("Entrer un nombre: "))
while nombre > 0:
  print(f"Décompte: {nombre}")
  nombre = nombre + 1

On peut effectuer les mêmes opérations avec les boucles `for` et `while`.

**Boucler de 0 à 9**
```python
for i in range(10):
  print(i)

i = 0
while i < 10:
  print(i)
  i = i + 1
```

**Calculer la somme des 50 premiers nombres**
```python
somme = 0
for i in range(1, 51):
  somme = somme + i
print(f"Somme: {somme}")

i = 1
somme = 0
while i <= 50:
  somme = somme + i
  i = i + 1
print(f"Somme: {somme}")
```

Cependant, pour les opérations dont le nombre d'itération est connu, la boucle `while` offre quelques inconvénients par rapport à la boucle `for`:
- Code plus difficile à lire
- Gestion manuelle de l'itérateur (variable `i`)
- Possibilité de boucle infinie

### Approximer $\pi$ avec la boucle `while`

En utilisant une série, il est possible d'approximer $\pi$ et d'arrêter automatiquement lorsque la précision désirée est obtenue. La formule utilisée sera
$$\pi ≈ \frac{4}{1} - \frac{4}{3} + \frac{4}{5} - \frac{4}{7} + ...$$
Étant donné qu'on ne sait pas d'avance combien d'itération il faudra, la boucle `while` est une bonne candidate.

In [ ]:
# TODO: Changer la tolérance pour obtenir un nombre de plus en plus près de PI.
# Note: Une tolérance de 1e-7 prend de 10 à 15 secondes à calculer.
pi = 0            # L'approximation de PI
nb_iterations = 0 # Le nombre d'itération nécessaire
TOLERANCE = 1e-2  # Le seuil où arrêter la boucle
variation = 4     # La variation obtenue
denominateur = 1  # Le dénominateur
est_plus = True   # Un drapeau pour déterminer l'addition ou la soustraction

while abs(variation) > TOLERANCE:
  # Calculer la variation, positive ou négative
  if est_plus:
    variation = 4 / denominateur
  else:
    variation = -4 / denominateur

  # Ajuster les variables
  nb_iterations = nb_iterations + 1
  denominateur = denominateur + 2
  est_plus = not est_plus
  pi = pi + variation

# Afficher l'approximation de PI
print(f"Nombre d'itérations: {nb_iterations}")
print(f"Approximation de PI: {pi}")

Nombre d'itérations: 201
Approximation de PI: 3.1465677471829556


### La boucle `while` et le `break`
L'instruction `break` permet de quitter les boucles `for` et `while`. Celle-ci permet d'ajouter une façon de quitter la boucle infinie d'une manière différente mais risquée. On peut forcer une boucle infinie avec `while True:`, tant et aussi longtemps que nous avons une condition de sortie avec l'instruction `break`.

In [ ]:
# TODO: Ajouter un compteur pour indiquer la quantité de nombre ajouté.
# TODO: Changer la précision partout pour n'afficher que 2 nombres après la virgule pour la somme.
somme = 0
while True:   # Boucle infinie!
  nombre = float(input(f"Somme de {somme}. Quel est le nombre à ajouter? Entrer 0 pour quitter. "))
  # Condition pour quitter
  if nombre == 0:
    break
  # Additionner la somme
  somme = somme + nombre

# Afficher la somme une fois la boucle terminée
print(f"La somme est {somme:.4f}.")

### 📄 Exercices
Faire les exercices de la section **La boucle `while`**

## 📘 Le module `random`

Le module `random` permet d'introduire une variante aléatoire à nos programmes. Il peut aider à simuler des événements pseudo-aléatoires.
```python
import random
print(random.random())              # Nombre aléatoire entre 0 et 1
print(random.randint(1, 6))         # Nombre aléatoire entre 1 et 6 (lancé de dé)
print(random.randrange(0, 20, 2))   # Nombre aléatoire entre 0 et 19 par saut de 2
```
Chaque fois que le code sera exécuté, de nouveaux nombres seront affichés.

In [ ]:
# TODO: Lancer le programme plusieurs fois pour s'apercevoir que les nombres sont toujours différents.
import random
print(random.random())              # Nombre aléatoire entre 0 et 1
print(random.randint(1, 6))         # Nombre aléatoire entre 1 et 6 (lancé de dé)
print(random.randrange(0, 20, 2))   # Nombre aléatoire entre 0 et 19 par saut de 2

Il est possible d'introduire un *seed* au module random pour définir un point de départ unique.
```python
import random
random.seed(150)
print(random.randint(1, 100))   # Affichera toujours 42

```
Ce point de départ servira à la génération de tous les nombres futurs et rendra le module `random` prévisible.

In [ ]:
# TODO: Lancer le programme plusieurs pour s'apercevoir que les valeurs sont identiques à chaque lancement.
# TODO: Changer le seed et comparer avec un voisin.
import random
random.seed(150)
print(random.randint(1, 100))
print(random.random())
print(random.randint(1, 6))
print(random.randrange(0, 20, 2))

En combinant les boucles, les conditionnelles et le module `random`, il est possible de simuler $n$ lancers de dé.

In [ ]:
import random

# Les variables de départ
n = 10000
qte_de_6 = 0

# Simuler n lancers de dé
for i in range(n):
  aleatoire = random.randint(1, 6)
  if aleatoire == 6:
    qte_de_6 = qte_de_6 + 1

# Afficher le résultat
print(f"Sur un total de {n} lancers, il y a eu {qte_de_6} lancé de 6, pour un pourcentage de {qte_de_6 / n * 100:.2f}%.")
print(f"La probabilité attendue est de 1/6, soit 16.66%.")

### 📄 Exercices
Faire les exercices de la section **Le module `random`**

## 📘 Les listes

Une liste permet d'emmagasiner plusieurs valeurs dans une seule variable. La liste est représentée par des crochets carrés `[` et `]` et les éléments sont séparés par une virgule `,`:
```python
liste = [2, 3, 4]
```
Les différents éléments peuvent ensuite être accédés en lecture ou en écrite à l'aide de leur position (*index*), en commençant par zéro:
```python
print(liste[0])  # Affiche 2
print(liste[1])  # Affiche 3
print(liste[2])  # Affiche 4
liste[1] = 8
print(liste)     # Affiche [2, 8, 4]
```
On peut aussi afficher les éléments en utilisant leur position inverse, en commençant par la fin:
```python
print(liste[-1])  # Affiche 4
print(liste[-2])  # Affiche 3
print(liste[-3])  # Affiche 2
```
Finalement, il est possible d'obtenir une plage de nombres (ex. les deux premiers):
```python
print(liste[0:2])  # Affiche [2, 3]
```

Une liste peut contenir des valeurs autres que des nombres:
```python
noms = ["Sébastien", "Olivier"]
```
On peut ajouter des éléments à l'aide de `append()` et les retirer à l'aide de `pop()`. Par défaut, `pop()` retourne et retire le dernier élément de la liste, sinon `pop(index)` retire l'élément à la position `index`.
```python
noms.append("Charles")
noms.append("Bob")
print(noms)          # Affiche ["Sébastien", "Olivier", "Charles", "Bob"]
retire = noms.pop()  # Enlève "Bob" de la liste
print(noms)          # Affiche ["Sébastien", "Olivier", "Charles"]
print(retire)        # Affiche "Bob"
```

In [ ]:
# TODO: Sans modifier le code existant:
#       - Enlever les personnes ne faisant pas parti de la classe
#       - Ajouter quelques personnes supplémentaires
#       - Afficher le 2e nom de la classe ainsi que l'avant dernier
#       - Afficher tous les noms entre le 2e et le 5e inclusivement
classe = []
classe.append("Sébastien")
classe.append("Bob")
classe.append("Alice")
classe.append("Olivier")
classe.append("Charlie")
classe.append("Daryl")
classe.append("Charles")
print(classe)
##
## Écrire votre code ci-dessous
classe.pop(5)
classe.pop(4)
classe.pop(2)
classe.pop(1)
print(classe)
classe.append("Stéphanie")
print(classe)
print(classe[1])
print(classe[-2])

['Sébastien', 'Bob', 'Alice', 'Olivier', 'Charlie', 'Daryl', 'Charles']
['Sébastien', 'Olivier', 'Charles']
['Sébastien', 'Olivier', 'Charles', 'Stéphanie']
Olivier
Charles


### Fonctions utiles avec les listes

#### `min()`, `max()` et `sum()`

Certaines statistiques peuvent être obtenues sur une liste de nombres.
```python
nombres = [2, 5, 9, 3, 5]
print(min(nombres))  # Affiche 2
print(max(nombres))  # Affiche 9
print(sum(nombres))  # Affiche 24
```

#### Taille d'une liste avec `len()`

La taille d'une liste peut être obtenue avec `len()`. Cette fonction est très pratique pour calculer rapidement une moyenne.

```python
nombres = [2, 5, 9, 3, 5]
somme = sum(nombres)
taille = len(nombres)
print(f"Somme: {somme}")             # Somme: 24
print(f"Taille: {taille}")           # Taille: 5
print(f"Moyenne: {somme / taille}")  # Moyenne: 4.8
```

#### Chercher un élément dans une liste

Pour savoir si un élément dans une liste existe déjà, on peut utiliser une conditionnelle `if` et le mot-clé `in` ou son inverse `not in`:
```python
noms = ["Sébastien", "Olivier", "Bob"]

if "Bob" in noms:
  print("Bob est là!")

if "Charles" not in noms:
  noms.append("Charles")
```

#### Ordonner une liste avec `sort()` et `sorted()`

Une liste peut être ordonnée de plusieurs manières avec les fonctions `sort()` et `sorted()`. Les deux fonctions sont presqu'identiques:
- `sort()` ordonne la liste directement dans la variable et ne retourne rien
- `sorted()` n'affecte pas la liste originale et retourne une nouvelle liste ordonnée

In [ ]:
# Sort()
nombres = [5, 3, 4, 1, 2]
nombres_ordonnees = nombres.sort()
print(nombres)
print(nombres_ordonnees)

[1, 2, 3, 4, 5]
None


In [ ]:
# Sorted()
nombres = [5, 3, 4, 1, 2]
nombres_ordonnees = sorted(nombres)
print(nombres)
print(nombres_ordonnees)

[5, 3, 4, 1, 2]
[1, 2, 3, 4, 5]


Une liste peut également être ordonnée à l'inverse en utilisant le paramètre `reverse=True` dans les fonctions `sort()` et `sorted()`.
```python
nombres = [5, 3, 4, 1, 2]
nombres.sort(reverse=True)
print(nombres)            # Affiche [5, 4, 3, 2, 1]

nombres_ordonnees = sorted(nombres, reverse=True)
print(nombres_ordonnees)  # Affiche [5, 4, 3, 2, 1]
```

In [ ]:
# TODO: Ajouter seulement les nombres pairs ou divisibles par 7
# TODO: Afficher la liste de manière décroissante
# TODO: Afficher la moyenne des nombres obtenus
import random
n = 20
liste = []
for i in range(n):
  nombre = random.randint(1, 100)

  if nombre % 2 == 0 or nombre % 7 == 0:
    liste.append(nombre)

print(liste)
liste.sort(reverse=True)
print(liste)

taille = len(liste)
somme = sum(liste)
moyenne = somme / taille
print(f"Moyenne: {moyenne}")

[50, 100, 10, 52, 78, 72, 6, 80, 10]
[100, 80, 78, 72, 52, 50, 10, 10, 6]
Moyenne: 50.888888888888886


#### Copier une liste avec `copy()`


Les listes sont des *objets* un peu spéciaux lorsque vient le temps de les réutiliser. D'habitude, le code
```python
x = 8
y = x
y = y + 2
print(x)   # Affiche 8
print(y)   # Affiche 10
```
siginfie de copier la valeur de `x` dans la variable `y` et nous obtenons deux variables distinctes. Cependant, pour les listes, le code suivant donnera un résultat un peu spécial:

In [ ]:
# TODO: Avant d'exécuter le code, déterminer le résultat
# TODO: Exécuter le code pour voir le résultat
liste_1 = [1, 2, 3, 4, 5]
liste_2 = liste_1
liste_2.append(6)

print(liste_1)
print(liste_2)

En effet, les deux variables *pointent* vers la même liste. Pour créer une copie, il est nécessaire d'utiliser la fonction `copy()` comme suit:
```python
liste_2 = copy(liste_1)
```

In [ ]:
# TODO: Corriger le code pour obtenir deux listes distinctes
liste_1 = [1, 2, 3, 4, 5]
liste_2 = liste_1
liste_2.append(6)

print(liste_1)
print(liste_2)

### Boucler à travers une liste

Il existe deux façons de boucler à travers une liste:
- La méthode *traditionnelle*, en passant à travers les index
- La méthode *python*, en bouclant directement dans la liste

```python
# Méthode traditionnelle
liste = ["Sébastien", "Olivier", "Charles"]
for i in range(len(liste)):
  nom = liste[i]
  print(nom)

# Méthode python
liste = ["Sébastien", "Olivier", "Charles"]
for nom in liste:
  print(nom)
```
À moins que l'index soit nécessaire au fonctionnement du programme, la méthode *python* est plus rapide et habituellement plus facile à lire.

In [ ]:
# TODO: Sans modifier le code existant:
#       - Mettre tous les nombres positifs de la liste dans la variable liste_positive
#       - Mettre tous les nombres négatifs de la liste dans la variable liste_negative
#       - Afficher la taille des deux listes
import random

liste = []
for i in range(1000):
  liste.append(random.randint(1, 1000))

##
# Écrire votre code ici
liste_positive = []
liste_negative = []

## 📘 Les tuples

Contrairement aux listes où l'ordre n'a habituellement pas d'importance, l'ordre des éléments d'un tuple est primordial. L'exemple le plus courant est une coordonnée $(x, y)$ sur un plan cartésien. Un tuple s'utilise comme une liste, mais se définir à l'aide de parenthèses au lieu de crochets carrés:
```python
coordonnee = (2, 3)
print(coordonnee)     # Affiche (2, 3)
print(coordonnee[0])  # Affiche 2
print(coordonnee[1])  # Affiche 3
```
La majorité des fonctions des listes sont utilisables avec les tuples, à l'exception de `sort()`. En effet, étant donné que l'ordre des éléments d'un tuple est important, celui-ci ne peut pas s'ordonner.

### 📄 Exercices
Faire les exercices de la section **Les listes et les tuples**

In [ ]:
coordonnee = (2, 3)
print(coordonnee)
print(coordonnee[0])  # Affiche 2
print(coordonnee[1])  # Affiche 3
print(min(coordonnee))
print(len(coordonnee))
print(sum(coordonnee))
sorted(coordonnee)

(2, 3)
2
3
2
2
5


[2, 3]

## 📘 Les dictionnaires

Les dictionnaires sont une autre façon d'organiser des données. Au lieu d'utiliser des index comme les listes et les tuples, ils utilisent des *clés*. On les définit à l'aide d'acolades `{` et `}`.
```python
permis = {
  "numero": "QC-12345",
  "nom": "Huneault",
  "prenom": "Sébastien",
  "yeux": "bleus",
  "age": 38,
  "est_adulte": True
}
```
❗Le dictionnaire a été mis sur plusieurs lignes pour faciliter la lecture.

Accéder à une valeur dans un dictionnaire est similaire à une liste: on utilise le nom de la variable, des crochets carrés et l'index:
```python
permis["numero"]
```

In [ ]:
# TODO: Changer les informations pour les vôtres et afficher votre nom, prénom et votre couleur de yeux.
permis = {
  "numero": "QC-12345",
  "nom": "Huneault",
  "prenom": "Sébastien",
  "yeux": "bleus",
  "age": 38,
  "est_adulte": True
}
print(permis["nom"])

### Boucler dans un dictionnaire

Comme pour une liste, il est possible de boucler à travers un dictionnaire. La boucle utilisera alors les clés comme itérateur.

In [ ]:
# TODO: À l'aide d'une boucle, afficher la clé et la valeur de chaque élément du dictionnaire.
permis = {
  "numero": "QC-12345",
  "nom": "Huneault",
  "prenom": "Sébastien",
  "yeux": "bleus",
  "age": 38,
  "est_adulte": True
}
for i in permis:
  print(i)

### Imbriquer une liste dans un dictionnaire

Les listes peuvent être intégrés dans un dictionnaire, au même titre que le texte, nombres et booléens:
```python
# Liste dans un dictionnaire
famille = {
  "pere": "Bob",
  "mere": "Alice",
  "enfants": [
    "Charlie",
    "Daryl",
    "Eve"
  ]
}
print(famille["enfants"][0])   # Afficher le premier enfant, Charlie
print(len(famille["enfants"])) # Afficher le nombre d'enfants, 3
```
Et un dictionnaire peut aussi être imbriqué dans une liste:
```python
# Dictionnaire dans une liste
epicerie = [
  { "item": "Pomme", "quantite": 3 },
  { "item": "Lait", "quantite": 1 },
  { "item": "Pain", "quantite": 2 }
]
print(len(epicerie))           # Afficher le nombre d'articles différents, 3
print(epicerie[2]["item"])     # Afficher le nom du 3e article, Pain
```

In [ ]:
# TODO: À l'aide d'une boucle, afficher combien d'item total sera acheté
# TODO: À l'aide d'une boucle, afficher le coût total d'achat de l'épicerie
epicerie = [
  { "item": "Pomme", "quantite": 3, "cout": 0.89 },
  { "item": "Lait", "quantite": 1, "cout": 6.99 },
  { "item": "Pain", "quantite": 2, "cout": 4.99 }
]
for article in epicerie:
  print(article)

### 📄 Exercices
Faire les exercices de la section **Les dictionnaires et tableaux imbriqués**